frame/contracts/src/wasm/mod.rs

```diff
@@ -248,64 +243,171 @@ impl<T: Config> PrefabWasmModule<T> {
 		Ok((store, memory, instance))
 	}
 
+	/// Put the module blob into storage.
+	///
+	/// Increments the reference count of the in-storage `WasmBlob`, if it already exists in
+	/// storage.
+	fn store_code(mut module: Self, instantiated: bool) -> DispatchResult {
+		let code_hash = &module.code_hash();
+		<CodeInfoOf<T>>::mutate(code_hash, |stored_code_info| {
+			match stored_code_info {
+				// Instantiate existing contract.
+				Some(stored_code_info) if instantiated => {
+					stored_code_info.refcount = stored_code_info.refcount.checked_add(1).expect(
+						"
+					 refcount is 64bit. Generating this overflow would require to store
+					 _at least_ 18 exabyte of data assuming that a contract consumes only
+					 one byte of data. Any node would run out of storage space before hitting
+					 this overflow;
+					 qed
+					",
+					);
+					Ok(())
+				},
+				// Contract code is already stored in storage. Nothing to be done here.
+				Some(_) => Ok(()),
+				// Upload a new contract code.
+				//
+				// We need to store the code and its code_info, and collect the deposit.
+				None => {
+					// This `None` case happens only in freshly uploaded modules. This means that
+					// the `owner` is always the origin of the current transaction.
+					T::Currency::reserve(&module.code_info.owner, module.code_info.deposit)
+						.map_err(|_| <Error<T>>::StorageDepositNotEnoughFunds)?;
+					module.code_info.refcount = if instantiated { 1 } else { 0 };
+					<PristineCode<T>>::insert(code_hash, module.code);
+					*stored_code_info = Some(module.code_info);
+					<Pallet<T>>::deposit_event(
+						vec![*code_hash],
+						Event::CodeStored { code_hash: *code_hash },
+					);
+					Ok(())
+				},
+			}
+		})
+	}
+
+	/// Try to remove code together with all associated information.
+	fn try_remove_code(origin: &T::AccountId, code_hash: CodeHash<T>) -> DispatchResult {
+		<CodeInfoOf<T>>::try_mutate_exists(&code_hash, |existing| {
+			if let Some(code_info) = existing {
+				ensure!(code_info.refcount == 0, <Error<T>>::CodeInUse);
+				ensure!(&code_info.owner == origin, BadOrigin); // TODO: what if origin is root?
```

Comment:
_(not related to this PR)_, just a thought: isn't it make sense to allow Root origin to remove contract?